x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,694 @@
+            String repositoryName,
+            BlobContainer delegatedBlobContainer,
+            CheckedSupplier<SingleUseKey, IOException> singleUseDEKSupplier,
+            CheckedFunction<String, SecretKey, IOException> getDEKById
+        ) {
+            super(path);
+            this.repositoryName = repositoryName;
+            final String rootPathElement = path.iterator().hasNext() ? path.iterator().next() : null;
+            if (DEK_ROOT_CONTAINER.equals(rootPathElement)) {
+                throw new RepositoryException(repositoryName, "Cannot descend into the DEK blob container " + path);
+            }
+            this.delegatedBlobContainer = delegatedBlobContainer;
+            this.singleUseDEKSupplier = singleUseDEKSupplier;
+            this.getDEKById = getDEKById;
+        }
+
+        /**
+         * Returns a new {@link InputStream} for the given {@code blobName} that can be used to read the contents of the blob.
+         * The returned {@code InputStream} transparently handles the decryption of the blob contents, by first working out
+         * the blob name of the associated DEK id, reading and decrypting the DEK (given the repository password, unless the DEK is
+         * already cached because it had been used for other blobs before), and lastly reading and decrypting the data blob,
+         * in a streaming fashion, by employing the {@link DecryptionPacketsInputStream}.
+         * The {@code DecryptionPacketsInputStream} does not return un-authenticated data.
+         *
+         * @param   blobName The name of the blob to get an {@link InputStream} for.
+         */
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            // This MIGHT require two concurrent readBlob connections if the DEK is not already in the cache and if the encrypted blob
+            // is large enough so that the underlying network library keeps the connection open after reading the prepended DEK ID.
+            // Arguably this is a problem only under lab conditions, when the storage service is saturated only by the first read
+            // connection of the pair, so that the second read connection (for the DEK) can not be fulfilled.
+            // In this case the second connection will time-out which will trigger the closing of the first one, therefore
+            // allowing other pair connections to complete.
+            // In this situation the restore process should slowly make headway, albeit under read-timeout exceptions
+            final InputStream encryptedDataInputStream = delegatedBlobContainer.readBlob(blobName);
+            try {
+                // read the DEK Id (fixed length) which is prepended to the encrypted blob
+                final byte[] dekIdBytes = new byte[DEK_ID_LENGTH];
+                final int bytesRead = Streams.readFully(encryptedDataInputStream, dekIdBytes);
```

Comment:
@tvernum regarding your inquiry about encrypted repo versioning:

Encrypted blobs have only the DEK Id (plaintext UUID) prepended, so the repo **currently doesn't support snapshots of multiple versions**. If we plan that future encrypted repos be backwards compatible with the current version, in a way that the new version be able to work with a repo of the current version, then we might be in trouble.  In this case, any given snapshot can technically contain blobs encrypted in multiple versions, so we have to rely on the encrypted blob itself to specify which version of the repo encryption "scheme" to use (when decrypting, i.e. snapshot restore).

I will be adding this item to the list from https://github.com/elastic/elasticsearch/issues/41910 . I concur we mustn't release before we have some way to accommodate possible future versions of the repo.

But, blob level versioning is tricky. I remember I coded serialisation of version headers tacked on encrypted blobs. Probably the version header must be authenticated, but not encrypted. All that is messy and I really appreciate the simplicity of just prepending a plaintext UUID to the encrypted blob (the current approach). 

I think prepending a "magic byte" to the UUID (which prepends the encrypted blobs) should be enough. Future versions might have a different value for the magic byte (which they might need to authenticate but not encrypt) and they might not even need one because they would infer the DEK by some other, more efficient,  means (eg. a snapshot-level dictionary of blob name to DEK content).